File: Java/commons/es-interfaces/src/main/java/org/es/AggregateRoot.java
package org.es;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;
import org.cqrs.Event;

public abstract class AggregateRoot {

    private final ConcurrentHashMap<Class, ConcurrentHashMap<Class, Consumer<Object>>> applyFunctions = new ConcurrentHashMap<>();

    protected void registerApply(Consumer<Object> applyFunction, Class messageType) {
        applyFunctions.putIfAbsent(this.getClass(), new ConcurrentHashMap<>());
        applyFunctions.get(this.getClass()).putIfAbsent(messageType, applyFunction);
    }

    private void doApply(Event event) {
        ConcurrentHashMap<Class, Consumer<Object>> aggregateApplier = applyFunctions.get(this.getClass());
        Consumer<Object> functionApplier = aggregateApplier.get(event.getClass());
        functionApplier.accept(event);
    }

    protected AggregateRoot() {
        initializeApply();
    }

    protected abstract void initializeApply();

    private final List<Event> changes = new ArrayList<>();
    private UUID id;

    public UUID getId() {
        return id;
    }

    public List<Event> getUncommittedChanges() {
        return changes;
    }

    public void MarkChangesAsCommitted() {
        changes.clear();
    }

    public void LoadsFromHistory(List<Event> history) {
        for (Event e : history) {
            applyChange(e, false);
        }
    }

    protected void applyChange(Event event) {
        applyChange(event, true);
    }

    private void applyChange(Event event, boolean isNew) {
        doApply(event);
        if (isNew) {
            changes.add(event);
        }
    }
}

File: Net/Commons/Es.Interfaces/AggregateRoot.cs
using Cqrs;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;

namespace Es.Interfaces
{
    public abstract class AggregateRoot
    {
        private static ConcurrentDictionary<Type, ConcurrentDictionary<Type, Action<Object>>> applyFunctions = new ConcurrentDictionary<Type, ConcurrentDictionary<Type, Action<object>>>();

        protected void RegisterApply(Action<Object> applyFunction, Type messageType)
        {
            applyFunctions[this.GetType()] = new ConcurrentDictionary<Type, Action<Object>>();
            applyFunctions[this.GetType()][messageType] = applyFunction;
        }

        private void DoApply(IEvent evt)
        {
            ConcurrentDictionary<Type, Action<Object>> aggregateApplier = applyFunctions[this.GetType()];
            Action<Object> functionApplier = aggregateApplier[evt.GetType()];
            functionApplier.Invoke(evt);
        }

        protected AggregateRoot()
        {
            InitializeApply();
        }

        protected abstract void InitializeApply();

        private List<IEvent> changes = new List<IEvent>();
        public Guid Id { get; protected set; }

        public List<IEvent> GetUncommittedChanges()
        {
            return changes;
        }

        public void MarkChangesAsCommitted()
        {
            changes.Clear();
        }

        public void LoadsFromHistory(List<IEvent> history)
        {
            foreach (IEvent e in history)
            {
                ApplyChange(e, false);
            }
        }

        private void ApplyChange(IEvent evt, bool isNew = true)
        {
            DoApply(evt);
            if (isNew)
            {
                changes.Add(evt);
            }
        }
    }
}
